### PR TITLE
[WIP] Interactive CLI output

### LINKED ISSUE #125: Display progress during the CLI run
To show that the app is not stuck it would be very nice to show some progress during the run.

We can show the current endpoint/method and number of executed tests

It could look like this:

```
GET:    /v1/users .......
POST:   /v1/users ...
```
And show the number of tests with dots (pytest style), maybe we can also show failures with "F", not sure.

To implement it we should not capture all stdout during `runner.execute` (`cli.__init:79`), but only hypothesis one + write the progress on each iteration of execution.

Maybe it will be easier to change `runner.execute_from_schema` into a generator so the runner itself will not know about another part that will write progress info to the terminal - I'd like to avoid adding this logic into the runner itself since it is relevant more to the CLI part.

Actually, if the control will be yielded upper from the runner, then that part could use `StatsCollector` to write the progress - this might need some restructuring in the runner. 

Something like this:
```
# cli/__init__.py
stats = StatsCollector()
for _ in runner.execute(..., stats):
    click.echo("some info from stats/runner")
```

Also if the runner will have an internal state, then it will be easier to report the current/next test.

On the other hand, the runner should be still easy to use, converting it to a generator shouldn't change the current API

### PR DESCRIPTION
**Pull request by svtkachenko**
Closes #125

schemathesis

### CODE REVIEW

File: src/schemathesis/runner/events.py [hunk 1]
```diff
@@ -24,6 +24,7 @@ class Initialized(ExecutionEvent):
 @attr.s(slots=True)
 class BeforeExecution(ExecutionEvent):
     endpoint: Endpoint = attr.ib()
+    totatl: int = attr.ib()
```
Comment (Stranger6667) [inline src/schemathesis/runner/events.py]: a typo. `totatl` -> `total`

File: src/schemathesis/runner/__init__.py [hunk 2]
```diff
@@ -89,9 +89,14 @@ def execute(  # pylint: disable=too-many-arguments
         hypothesis_options=hypothesis_options,
         loader=loader,
     )
-    all_events = list(generator)
-    finished = all_events[-1]
-    return finished.statistic
+    for r in generator:
```
Comment (Stranger6667) [inline src/schemathesis/runner/__init__.py]: Do we need it as a generator? This one shouldn't be used by CLI, CLI should utilize `execute_as_generator`

File: src/schemathesis/schemas.py [hunk 3]
```diff
@@ -55,6 +55,10 @@ def resolver(self) -> jsonschema.RefResolver:
             self._resolver = jsonschema.RefResolver("", self.raw_schema)
         return self._resolver
 
+    @property
+    def test_cases_count(self):
```
Comment (Stranger6667) [inline src/schemathesis/schemas.py]: let's rename it to `endpoints_count`

File: src/schemathesis/cli/output.py [hunk 4]
```diff
@@ -62,3 +69,65 @@ def pretty_print_stats(stats: runner.StatsCollector, hypothesis_output: Optional
                     click.style(verdict, fg=color, bold=True),
                 )
             )
+
+
+def pct(position: int, length: int) -> str:
```
[the remark this answers is not in the document]
Comment (svtkachenko) [reply src/schemathesis/cli/output.py]: pct, an abbreviation for "%"

File: src/schemathesis/cli/output.py [hunk 5]
```diff
@@ -62,3 +69,65 @@ def pretty_print_stats(stats: runner.StatsCollector, hypothesis_output: Optional
...
+    return f"[{position * 100 // length}%]"
+
+
+def pretty_print_test_progress(results_generator: Generator[events.ExecutionEvent, None, None]) -> StatsCollector:
+    # Starting progress position
+    position = 0
+    errors = 0
+    # Terminal size and fields position
+    columns = shutil.get_terminal_size((100, 50)).columns
```
[the remark this answers is not in the document]
Comment (svtkachenko) [reply src/schemathesis/cli/output.py]: Same as in `print_in_section`

[the remark this answers is not in the document]
Comment (svtkachenko) [reply src/schemathesis/cli/output.py]: To conform other output defaults.

File: src/schemathesis/cli/output.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Stranger6667) [reply src/schemathesis/cli/output.py]: I'd rather avoid abbreviations to reduce a cognitive load when reading the source code

[the remark this answers is not in the document]
Comment (Stranger6667) [reply src/schemathesis/cli/output.py]: :D

File: src/schemathesis/cli/output.py [hunk 6]
```diff
@@ -62,3 +69,65 @@ def pretty_print_stats(stats: runner.StatsCollector, hypothesis_output: Optional
...
+            template = f"{{:{col3_len}}} "
+            if new_errors > errors:
+                click.echo(template.format("F"), nl=False)
+            else:
+                click.echo(template.format("."), nl=False)
+            col4_len = int(columns) - col1_len - col2_len - col3_len - 10
+            click.echo(f"{{:>{col4_len}}}".format(percentage(position, init.schema.endpoints_count)))  # type: ignore
+        if isinstance(result, events.Finished):
+            return result.statistic
+    return StatsCollector()
```
Comment (Stranger6667) [inline src/schemathesis/cli/output.py]: I see no point to return an empty StatsCollector. Probably the return type should be `None`

Comment (Stranger6667) [reply src/schemathesis/cli/output.py]: Hm, please, ignore the comment above

**Pull request merged**